### PR TITLE
DEPS: Upgrade Deps for Python 3.10

### LINKED ISSUE #40307: Use oldest-supported-numpy in pyproject.toml
We should be able to replace the build requires section of our pyproject.toml with https://pypi.org/project/oldest-supported-numpy/

### LINKED ISSUE #43715: BUG: pandas doesn't install on python 3.10
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
pip install pandas (installs 1.3.3)
```

### Issue Description

I tried to install on  3.10.0rc2 
```
ultiarray\scalartypes.obj
    scalartypes.c
    numpy\core\include\numpy/npy_3kcompat.h(227): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\common.h(239): warning C4244: 'return': conversion from 'npy_uintp' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(425): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(426): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(1899): warning C4244: 'function': conversion from 'npy_double' to 'npy_float', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(1902): warning C4244: 'function': conversion from 'npy_double' to 'npy_float', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2577): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2561): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2577): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
    numpy\core\src\multiarray\scalartypes.c.src(2967): error C2440: 'function': cannot convert from 'double' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2967): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2967): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(2976): error C2440: 'function': cannot convert from 'double' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2976): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2975): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(2982): error C2440: 'function': cannot convert from 'double' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2982): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2981): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(2967): error C2440: 'function': cannot convert from 'npy_longdouble' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2967): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2967): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(2976): error C2440: 'function': cannot convert from 'npy_longdouble' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2976): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2975): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(2982): error C2440: 'function': cannot convert from 'npy_longdouble' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2982): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2981): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(2997): error C2440: 'function': cannot convert from 'double' to 'PyObject *'
    numpy\core\src\multiarray\scalartypes.c.src(2997): warning C4024: '_Py_HashDouble': different types for formal and actual parameter 1
    numpy\core\src\multiarray\scalartypes.c.src(2997): error C2198: '_Py_HashDouble': too few arguments for call
    numpy\core\src\multiarray\scalartypes.c.src(3913): warning C4244: 'return': conversion from 'npy_intp' to 'int', possible loss of data
```

### Expected Behavior

Works

### Installed Versions

Not relevant

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] xref #43715
- [ ] closes #40307
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
i think we have an issue about this but

https://github.com/scipy/oldest-supported-numpy

is there a reason we don't just use this?

File: pyproject.toml [hunk 1]
```diff
@@ -4,18 +4,25 @@
 requires = [
     "setuptools>=51.0.0",
     "wheel",
-    "Cython>=0.29.21,<3",  # Note: sync with setup.py
+    "Cython>=0.29.24,<3",  # Note: sync with setup.py
     # Numpy requirements for different OS/architectures
-    # Copied from https://github.com/scipy/scipy/blob/master/pyproject.toml (which is also licensed under BSD)
+    # Adapted from https://github.com/scipy/scipy/blob/master/pyproject.toml (which is also licensed under BSD)
     "numpy==1.17.3; python_version=='3.7' and (platform_machine!='arm64' or platform_system!='Darwin') and platform_machine!='aarch64'",
```
Comment (fangchenli) [inline pyproject.toml:10]: can remove this.

**Review by fangchenli — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (fangchenli) [inline pyproject.toml:10]: we can remove this.

**Review by fangchenli — CHANGES_REQUESTED** [review]

**Review by fangchenli — CHANGES_REQUESTED** [review]
we can remove the line about for python 3.7

**Comment by fangchenli** [comment]
> is there a reason we don't just use this?

There is an open drift #43146

**Comment by jreback** [comment]
can you add a note about the cython change in 1.4 whatsnew

**Comment by jreback** [comment]
docs error should be fixed on master if you can rebase

File: setup.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def is_platform_mac():
     return sys.platform == "darwin"
 
 
-min_cython_ver = "0.29.21"  # note: sync with pyproject.toml
+min_cython_ver = "0.29.24"  # note: sync with pyproject.toml
```
Comment (jreback) [inline setup.py:40]: do we have any docs that list the cython requirement and need updating?

Comment (lithomas1) [reply setup.py:40]: Just the whatsnew. I'll try to update this PR soonish.

Comment (jreback) [reply setup.py:40]: can we get away with not updating cython? if so then i think we can backport this

Comment (lithomas1) [reply setup.py:40]: Nope.

File: .github/workflows/sdist.yml [hunk 3]
```diff
@@ -23,7 +23,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8", "3.9"]
+        python-version: ["3.8", "3.9", "3.10-dev"]
```
Comment (fangchenli) [inline .github/workflows/sdist.yml:26]: Python 3.10 is released yesterday. I wonder if GitHub Actions would have it soon.

**Review by jreback — APPROVED** [review]
lgtm ping on green

**Review by fangchenli — APPROVED** [review]

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/43729/checks?check_run_id=3809383668 :-<

**Comment by lithomas1** [comment]
Note that this will fail right now until we can pick up the newly uploaded Python 3.10 uploaded by conda-forge. I expect this to pass in the next few days.

**Comment by jreback** [comment]
ok fair enough

pls rebase when it's ready then

File: setup.cfg [hunk 4]
```diff
@@ -30,7 +31,10 @@ project_urls =
 [options]
 packages = find:
 install_requires =
-    numpy>=1.18.5
+    numpy>=1.18.5; platform_machine!='aarch64' and platform_machine!='arm64' and python_version<'3.10'
+    numpy>=1.19.2; platform_machine=='aarch64'
+    numpy>=1.20.0; platform_machine=='arm64'
+    numpy>=1.21.0; python_version=='3.10'
```
Comment (henryiii) [inline setup.cfg:37]: Is there a reason to list these out? If you are on Python 3.10, there simply won't be a version of Python older than 1.21, if you are on a special arch, same thing; so why limit it here? Also this exact string won't work when Python 3.11 comes along.

**Comment by lithomas1** [comment]
Okay, this looks working and the test from #43146 has been added. 
@henryiii Can you take one last look? Thanks for all the help so far.

**Comment by simonjayhawkins** [comment]
@lithomas1 needs manual backport when you can

**Comment by simonjayhawkins** [comment]
I think asv_bench/asv.conf.json needs to be updated too? (Are the benchmarks running?) Any why is environment.yml and requirements-dev.txt not updated?

**Pull request merged**